1766: Unify date picker components

andrew8er opened the pull request:
### Short Description

Replaces `<TextField>` components in `DateForm` with `<CustomDatePicker>`.

### Proposed Changes

- Updates the package `@mui/x-date-pickers` to 7.27.3.
- Replaces the `<TextField>` component in `DateForm` with `<CustomDatePicker>` components. 
- Uses `enableAccessibleFieldDOMStructure` for `<CustomDatePicker>`, which will be the [future default](enableAccessibleFieldDOMStructure) for MUI components.
- Uses the proper `format` for `<CustomDatePicker>`, fixing the text input. The format will also be displayed and having the month fields in lower case looks slightly misplaced, but I think this is tolerable. Could be fixed with some more CSS.
- Remove some box shadows that looked misplaced.
- Make all event handlers optional in `<CustomDatePicker>`.

### Side Effects

None.

### Testing

- Start backend and frontend
- Visit http://localhost:3000/beantragen and look at date form fields

### Resolved Issues

<!-- List all issues which should be closed when this PR is merged. -->

Fixes: #1766

f1sh1918 commented on administration/src/mui-modules/application/primitive-inputs/DateForm.tsx:
```diff
@@ -40,28 +43,29 @@ const DateForm: Form<State, ValidatedInput, AdditionalProps, Options> = {
-          min: '1900-01-01',
-        }}
+      <CustomDatePicker
+        date={null}
```
🔧 it looks like the `DesktopDatePicker` from mui also accepts `undefined`
So you could remove this prop here and adjust the `customDatePickerProps`
`date?: Date`

f1sh1918 reviewed (changes requested):
Looks quite good so far except of the styling doesn't fit to the forms of the particular projects. Here are some examples

Koblenz & Nürnberg
https://staging.koblenz.sozialpass.app/erstellen

Bayern
https://staging.bayern.ehrenamtskarte.app/beantragen

michael-markl commented:
Please check whether the input form actually works on mobile devices (at least for Safari, Chrome & Firefox) (I am just adding this as the name `DesktopDatePicker` suggests otherwise :) )

f1sh1918 commented:
> Please check whether the input form actually works on mobile devices (at least for Safari, Chrome & Firefox) (I am just adding this as the name `DesktopDatePicker` suggests otherwise :) )

We already use this component for koblenz and it was tested.
Don't panic @michael-markl :)

michael-markl commented:
Well, it doesn't work well on my Firefox for Android 😬

steffenkleinle commented:
> Well, it doesn't work well on my Firefox for Android 😬

Same for me actually.

f1sh1918 commented:
Ok i expected that it was already tested on different Browsers before. For chrome it works like a charme.
That's now really bad before the koblenz launch :/

michael-markl commented on administration/src/mui-modules/application/primitive-inputs/DateForm.tsx:
```diff
@@ -40,28 +41,27 @@ const DateForm: Form<State, ValidatedInput, AdditionalProps, Options> = {
-          min: '1900-01-01',
-        }}
+      <CustomDatePicker
+        date={null}
```
We have to use controlled values here, as the state of the application form is persisted in local storage.

andrew8er replied:
That was committed by accident, sorry.

connium commented:
I have checked several websites that offer a date picker. On most of them, the user is shown a date picker and cannot enter the date value via a text field. The native `<input type="date">` element is also implemented this way.

Since Firefox seems to be the only browser where the text input does not work at the moment, I suggest to use the `DatePicker` (responsive variant) for Firefox and keep the `DesktopDatePicker` for all other browsers.

michael-markl commented on administration/src/bp-modules/components/CustomDatePicker.tsx:
```diff
@@ -1,51 +1,69 @@
+      disabled={disabled}
+      label={label}
       views={['year', 'month', 'day']}
       value={date}
```
I think it would also be nice to add the prop `openTo="year"` as selecting a year is probably necessary for every of our date fields.

andrew8er commented:
@michael-markl 
 > Well, it doesn't work well on my Firefox for Android 😬

This should work now.

---

@connium 
>Since Firefox seems to be the only browser where the text input does not work at the moment, I suggest to use the DatePicker (responsive variant) for Firefox and keep the DesktopDatePicker for all other browsers. This should be fine for the users, as they are familiar with it from most sites.

The text input should work now as intended. We could however think about using a `DatePicker` here. The difference is explained [here](https://mui.com/x/react-date-pickers/base-concepts/#responsiveness)

f1sh1918 commented on administration/src/bp-modules/components/CustomDatePicker.tsx:
```diff
@@ -1,55 +1,75 @@
+      openTo='year'
       value={date}
-      format='dd.MM.yyyy'
+      format='dd.mm.yyyy'
```
this breaks the date input for me. if i type a year. month will change to `00`

andrew8er replied:
Mh, works for me both in FF and Chrome, both mobile and desktop. Can you give me the exact steps to reproduce?

andrew8er commented:
#### Month issue on FF Mobile
So, currently I can reproduce the issue as follows:

- Click into a `<CustomDatePicker>` text field, select the days.
- Enter some days, focus will change to months
- Try to enter months

On Firefox Mobile, when entering months with a leading zero, the input fails, months with a leading `1` however will work. Also month input works again, when switching to year input and back to month input. Firefox Desktop, as well as Chrome Desktop/Mobile and Safari Desktop are unaffected.

- The corresponding issue in `mui-x` is most likely this: https://github.com/mui/mui-x/issues/13854
- The underlying issue is actually a Firefox bug, tracked here: https://bugzilla.mozilla.org/show_bug.cgi?id=1947143

Possible solutions:
1. Scrap the text input, only use a popover, i.e. what `<MobileDatePicker>` provides.
2. As @connium suggested, we could provide an extra version for FF users. I'm not a big fan of this, since this would make the behavior dependant on the user's browser and complicates user support and documentation quiet a bit.
3. Wait for the FF issue to be fixed and ignore the problem for now.
    - It is still possible to enter a date with the popover, the behavior is annoying though.
    - FF Mobile usage is probably minuscule.
    - The bug is already triaged and assigned to someone, but it will take some month for a fix to land in the release version.

#### Styling
I got the style pretty close to the current style, but the component height in the case for KoblenzPass is still slightly too high. Fixing this would require a lot of extra CSS however, because the component makes some assumptions about it's height when displaying the label (as seen on the Ehrenamtskarte Bayern form, not to be confused with the placeholder). The date picker's text field has its own prop `size` where the components size can be controlled in two steps.

The better solution IMHO would be to use MUI for all the input fields in KoblenzPass, since the current input components come from `blueprintjs`.

f1sh1918 commented:
I'm okay with 3. since its rarely that people use firefox mobile and you can even use the datepicker in that case.

There are some small styling changes i would suggest f.e. the textfield `variant : filled` (for bavarian application process) and the `backgroundColor: white` is for that variant needed

I even couldn't find a way to reduce the paddings on quick testing but probably used the wrong class/prop for styling.

<img width="1137" alt="image" src="https://github.com/user-attachments/assets/dc130e9f-2945-4bd9-aa94-c6316a9f7981" />

We already created a ticket for switching to mui also for koblenz so no worry in the slight height difference

f1sh1918 commented on administration/src/bp-modules/components/CustomDatePicker.tsx:
```diff
@@ -1,55 +1,81 @@
           error: !isValid,
           spellCheck: false,
           onBlur,
+          style: textFieldStyle,
```
pass here a variant.
` variant: 'filled',` for bavaria and `outlined` for koblenz

f1sh1918 commented on administration/src/bp-modules/components/CustomDatePicker.tsx:
```diff
@@ -1,55 +1,81 @@
+              : undefined,
+            '.MuiPickersInputBase-root': {
+              fontSize: viewportSmall ? '16px' : '14px',
+              borderRadius: '2px',
```
add here `backgroundColor: 'white'` that is required for the `filled` variant to align the background styling

andrew8er replied:
I added some code to accommodate all kind of styles to match the requested style (see the last commit in this PR). However, MUI is clearly not designed for this. I think at some point we need to setup a way to have one theme for all MUI components per customer.

andrew8er commented:
The Material style EhrenamtsKarte Bayern is pretty dated now and not natively supported by MUI date pickers. I got it to work, but it is not pretty and we should think about updating the style, since its UX considered [inferior](https://medium.com/google-design/the-evolution-of-material-designs-text-fields-603688b3fe03) by the Google Design research team. Maybe as part of #1779.

f1sh1918 reviewed (approved):
Good work 👍 
Looks good

Just two small issues:
1. a11y issue that seem to be related to `enableAccessibleFieldDOMStructure`
The form field can not be escaped via tab if it has a value and the clear icon is shown
2. The second nit-picky one. Since you added this great deep merge for `textFieldSlotProps` could you please also pass the prop from the `BirthdayExtension` to adjust the height of the field.
```js
        textFieldSlotProps={{
          sx: {
            '.MuiPickersSectionList-root': {
              padding: '5px 0',
            },
          },
        }}
```
I think its a quite easy fix to adjust the height to look 99% equal :)

andrew8er commented:
> a11y issue that seem to be related to `enableAccessibleFieldDOMStructure` The form field can not be escaped via tab if it has a value and the clear icon is shown

I think this is a bug in the component itself, though I haven't found an issue. I tried a few options, but haven't found a good solution yet. We could use to disable keyboard access to both buttons (with `tabIndex: -1`) , but that is not making things better. The popover also has some keyboard navigation issues, for example, I cannot navigate beyond the first row of months.

seluianova commented:
:upside_down_face: nitpick, but the font in the birthday field is smaller then in other fields
![image](https://github.com/user-attachments/assets/30ce8451-940e-4cfa-8045-fb5181f2dc74)

seluianova commented:
we also have date in this form:
![image](https://github.com/user-attachments/assets/c38ef621-100b-44bb-8c8d-5c6f95bf8f6e)
do we want to adjust it as well?

seluianova reviewed (commented):
The form allows me to submit an invalid year, such as 0011, if I enter it from the keyboard
![image](https://github.com/user-attachments/assets/8a176424-b107-497a-ad31-4e731130930f)

andrew8er commented:
@seluianova 
> nitpick, but the font in the birthday field is smaller then in other fields

This was a style from the original component, but I agree, it looked out of place. Removed…

>  we also have date in this form:

I replaced the missing `<TextField>`s with `<CustomDatePicker>`, hope I caught everything now.

> The form allows me to submit an invalid year, such as 0011, if I enter it from the keyboard

We currently do not have any minimum date defined here, so this is technically correct. The original component provided a year picker starting from 1900 though. I think this would be a reasonable minimum date.

seluianova commented:
> We currently do not have any minimum date defined here

Yes, but in the current version of the form (main branch) it is not possible to submit a year such as 0011.
So I would treat this as a minor regression bug.
However if it's tricky to fix, I think we can create a separate issue for that, since it's not critical.

seluianova commented:
> We currently do not have any minimum date defined here

Yes, but in the current version of the form (main branch) it is not possible to submit a year such as 0011.
So I would treat this as a minor regression bug.
However if it's tricky to fix, I think we can create a separate issue for that, since it's not critical.

andrew8er commented:
The `<CustomDatePicker>` has a default `minDate` of 1900-01-01, but does not enforce it when entering the date via the text field, so I added another validation check.

seluianova commented:
:upside_down_face: When the birthday field is blank AND untouched, if I click “Nächster Schritt” without selecting a date, the error “Feld ist erforderlich.” appears in gray:
![image](https://github.com/user-attachments/assets/6c151fd1-2c9c-491e-ba09-bfb9b6caebea)
Expected: it should be red.
![image](https://github.com/user-attachments/assets/87007def-e967-44fa-b4d0-fea22d1c6c55)

Sorry, I can't resist testing tricky cases when I review PRs 😀
But I think it's fine to move this "issue" to a separate ticket since it's not a blocking issue, up to you

seluianova commented:
In the card creation form (/cards/add) Ablaufdatum is outlined in red, even if the value is valid:
![image](https://github.com/user-attachments/assets/dd5f488d-7e26-4593-b503-395c39fbf873)

seluianova commented:
:upside_down_face: another nit-pick, probably might be ignored, but these fields look unnececarily long and the space on the right looks a bit odd:
![image](https://github.com/user-attachments/assets/59f5a8c0-f81b-4411-afa0-236a583aaf5e)
This is how they looked like before:
![image](https://github.com/user-attachments/assets/08468a29-5c77-4b31-b2fd-d2da545dae56)

seluianova commented:
:upside_down_face: Nürnberg project --> cards creation form (/cards/add):
As a tooltip, there was a validation error if the date was earlier than the min date.
![image](https://github.com/user-attachments/assets/48706a66-e08b-4b1a-88a5-1b584506e42a)
Now this message is not displayed.

andrew8er commented:
#### birthday field
> When the birthday field is blank AND untouched, if I click “Nächster Schritt” without selecting a date, the error “Feld ist erforderlich.” appears in gray:

This does not happen for me with either FF or Chrome. Which browser are you using? Are the other fields filled? When clicking on the "Nächster Schritt" button, I get a notice to fill the form fields, nothing else. When I click on the birthday field, I get a red notice, as expected.

#### Misaligned fields
> Fields are missaligned here:

The misalignment stems from a lot of style customization for date picker's input field to match the current style. With #1779 we can ditch all those custom styles and this should be fixed automatically (assuming the MUI textfields are implemented coherently).

#### card creation date validation
> In the card creation form (/cards/add) Ablaufdatum is outlined in red, even if the value is valid:

Good catch! Fixed

#### date field icon padding
> another nit-pick, probably might be ignored for now, but these fields look unnececarily long and the space on the right looks a bit odd:

I noticed that too, and already tried to poke around the date picker's CSS but could not find a solution in a reasonable time. We already mess around the MUI date pickers CSS more than the the developers probably intended, but when implementing #1779, these should look normal again. If you find a quick'n'dirty solution, feel free to add a commit however!

#### missing tooltip
> 🙃 Nürnberg project --> cards creation form (/cards/add):
As a tooltip, there was a validation error if the date was earlier than the min date.

This tooltip here was automatically generated by the MUI `<TextField>`s. I think tooltips have a bad UX, especially for mobile. We generally use `helperText`s for displaying an error message beneath forms element. I could add this, but we would need appropriate translations.

f1sh1918 commented on administration/src/bp-modules/components/CustomDatePicker.tsx:
```diff
@@ -1,62 +1,88 @@
+        textFieldSlotProps
+      ),
+    }}
+    enableAccessibleFieldDOMStructure
```
can you please enable this only for viewport small @andrew8er 
Not be able to escape the field with keyboard is bad a11y

andrew8er replied:
MUI switches between the desktop and mobile variant based on the `@media (pointer: fine)` query, as explained [here](https://mui.com/x/react-date-pickers/date-picker/#available-components). I think it would be more consistent to use this query here also then.

f1sh1918 replied:
I'm open for any solutions as long as I can leave the field with keyboard :)

seluianova commented:
> This does not happen for me with either FF or Chrome. Which browser are you using? Are the other fields filled?

@andrew8er yes, all mandatory fields before Birthday should be filled and valid, otherwise when the button is clicked, it validates only the first invalid field, and the rest are not validated.
I use Chrome.

seluianova commented:
> This does not happen for me with either FF or Chrome. Which browser are you using? Are the other fields filled?

@andrew8er yes, all mandatory fields before Birthday should be filled and valid, otherwise validation of the birthday field will not trigger.
I use Chrome.

michael-markl commented on administration/src/mui-modules/application/primitive-inputs/DateForm.tsx:
```diff
@@ -41,28 +50,46 @@ const DateForm: Form<State, ValidatedInput, AdditionalProps, Options> = {
+        }}
+        textFieldHelperText={(showAllErrors || touched) && isInvalid ? validationResult.message : undefined}
+        textFieldSlotProps={{
+          variant: 'filled',
+          style: { marginTop: '12px', minWidth },
+          sx: {
+            '.MuiPickersInputBase-root': {
+              backgroundColor: 'transparent',
+            },
+            '.MuiPickersSectionList-root': {
+              paddingLeft: 0,
+            },
+            '.MuiInputLabel-root': {
+              transform: 'translate(0px, 13px) scale(1)',
+            },
+            '.MuiInputLabel-shrink': {
+              transform: 'translate(0px, -1.5px) scale(0.75)',
+            },
+            '.MuiFormHelperText-root': {
+              marginLeft: 0,
+            },
+          },
```
```suggestion
          variant: 'standard',
          style: { margin: '4px 0', minWidth },
          sx: {
            'input': { paddingTop: '4px' },
            '.MuiInputAdornment-root': { transform: 'translateX(-8px)' }
          }
```

I tried locally to use the standard variant, and it seems to work better than filled, with only some minor adjustments.

Feel free to ignore this comment though.

andrew8er replied:
Agreed, this is better.

f1sh1918 commented:
Just found two small issue in the statistics button bar
1. There is a padding-right that shouldn't be there after the icon
2. Labels are not vertically center

<img width="1011" alt="image" src="https://github.com/user-attachments/assets/bcacda1d-589b-41bf-83d0-279dbddf1e95" />

The rest looks and works good afaics :)

andrew8er commented:
>    There is a padding-right that shouldn't be there after the icon

This came from a overly unpecific nested style in the `<StickyBottomBar>`, took me an hour to find 🤯 

>    Labels are not vertically center

I removed all blueprint components and replaced them with MUI components, look good IMHO now.

michael-markl reviewed (approved):
Nice work! Looks good to me. (I tested Bayern application form, statistics date form, cards creation date form, and Koblenz self-portal form)
